File: tools/sync_all_branches.py
import subprocess
import sys

# SigmaOS Branch Uniformity & Synchronization Engine (S-BUSE)
# Automates absolute architectural parity across all 12 branches of the sovereign lattice.

def run_git(args):
    print(f"[RUN] git {' '.join(args)}")
    res = subprocess.run(["git"] + args, capture_output=True, text=True)
    if res.returncode != 0:
        print(f"[ERROR] git {' '.join(args)} failed:\n{res.stderr}")
        return False
    return True

def sync_branches():
    branches = [
        "release/standalone",
        "release/rtos",
        "release/mobile",
        "release/microkernel",
        "release/dual-boot",
        "release/distributed",
        "release/cloud",
        "release/browser",
        "release/app",
        "performance-optimized",
        "gh-pages"
    ]

    print("=========================================================================")
    print("SIGMAOS: BRANCH UNIFORMITY & SYNCHRONIZATION ENGINE [ACTIVE]")
    print("=========================================================================")

    # Ensure we start on main
    if not run_git(["checkout", "main"]):
        print("[FATAL] Could not checkout main. Aborting.")
        sys.exit(1)

    for branch in branches:
        print(f"\n[*] Syncing branch: {branch} -> Uniformity with main...")
        
        # Checkout the target branch
        if not run_git(["checkout", branch]):
            # Try creating it from main if checkout fails
            print(f"[!] Branch '{branch}' could not be checked out. Attempting to create it from main...")
            if not run_git(["checkout", "-b", branch]):
                print(f"[ERROR] Failed to switch to or create branch '{branch}'. Skipping.")
                continue

        # Hard reset to main to guarantee bit-perfect uniformity
        if not run_git(["reset", "--hard", "main"]):
            print(f"[ERROR] Failed to reset '{branch}' to main. Skipping.")
            continue

        # Force push the update to remote
        if not run_git(["push", "origin", branch, "--force"]):
            print(f"[ERROR] Failed to push branch '{branch}' to remote.")

    # Always return to main
    print("\n[*] Returning to main branch...")
    run_git(["checkout", "main"])
    print("\n=========================================================================")
    print("SIGMAOS: BRANCH SYNCHRONIZATION COMPLETE. PARITY ACHIEVED ACROSS ALL 12 BRANCHES.")
    print("=========================================================================")

if __name__ == "__main__":
    sync_branches()

File: tools/sync_all_branches.js
import { spawnSync } from 'child_process';

const branches = [
    "release/standalone",
    "release/rtos",
    "release/mobile",
    "release/microkernel",
    "release/dual-boot",
    "release/distributed",
    "release/cloud",
    "release/browser",
    "release/app",
    "performance-optimized",
    "gh-pages"
];

function runGit(args) {
    console.log(`[RUN] git ${args.join(' ')}`);
    try {
        const res = spawnSync("git", args, { stdio: 'pipe', encoding: 'utf-8' });
        if (res.status !== 0) {
            console.error(`[ERROR] failed:\n${res.stderr || res.stdout}`);
            return false;
        }
        return true;
    } catch (e) {
        console.error(`[ERROR] failed:\n${e.message}`);
        return false;
    }
}

console.log("=========================================================================");
console.log("SIGMAOS: NODE.JS BRANCH UNIFORMITY & SYNCHRONIZATION ENGINE [ACTIVE]");
console.log("=========================================================================");

// Ensure we start on main
if (!runGit(["checkout", "main"])) {
    console.error("[FATAL] Could not checkout main. Aborting.");
    process.exit(1);
}

for (const branch of branches) {
    console.log(`\n[*] Syncing branch: ${branch} -> Uniformity with main...`);
    
    // Checkout target branch
    if (!runGit(["checkout", branch])) {
        console.log(`[!] Branch '${branch}' could not be checked out. Attempting to create it from main...`);
        if (!runGit(["checkout", "-b", branch])) {
            console.error(`[ERROR] Failed to switch to or create branch '${branch}'. Skipping.`);
            continue;
        }
    }

    // Merge main into target branch with merge strategy -X theirs
    console.log(`[*] Merging core improvements from main into ${branch}...`);
    if (!runGit(["merge", "main", "-m", "Sync: Merge latest core improvements from main"])) {
        console.log(`[!] Conflict detected during merge. Resolving by preferring core improvements from main...`);
        runGit(["merge", "--abort"]);
        // Use -X theirs to auto-resolve in favor of main's updates
        if (!runGit(["merge", "-X", "theirs", "main", "-m", "Sync: Merge latest core improvements from main (conflict resolved)"])) {
            console.error(`[ERROR] Failed to merge main into branch '${branch}'. Skipping push.`);
            continue;
        }
    }

    // Push the synchronized branch to remote
    if (!runGit(["push", "origin", branch])) {
        console.log(`[!] Standard push failed. Attempting force push for parity...`);
        if (!runGit(["push", "origin", branch, "--force"])) {
            console.error(`[ERROR] Failed to push branch '${branch}' to remote.`);
        }
    }
}

// Always return to main
console.log("\n[*] Returning to main branch...");
runGit(["checkout", "main"]);
console.log("\n=========================================================================");
console.log("SIGMAOS: BRANCH SYNCHRONIZATION COMPLETE. PARITY ACHIEVED ACROSS ALL 12 BRANCHES.");
console.log("=========================================================================");
